bin/varnishd/cache/cache_hash.c

```diff
@@ -672,82 +672,101 @@ unsigned
 HSH_Purge(struct worker *wrk, struct objhead *oh, vtim_real ttl_now,
     vtim_dur ttl, vtim_dur grace, vtim_dur keep)
 {
-	struct objcore *oc, **ocp;
-	unsigned spc, ospc, nobj, n, n_tot = 0;
-	int more = 0;
+	struct objcore *oc, *oc_nows[2], **ocp;
+	unsigned i, j, n, n_max, total = 0;
 
 	CHECK_OBJ_NOTNULL(wrk, WORKER_MAGIC);
 	CHECK_OBJ_NOTNULL(oh, OBJHEAD_MAGIC);
-	ospc = WS_ReserveAll(wrk->aws);
-	assert(ospc >= sizeof *ocp);
-	/*
-	 * Because of "soft" purges, there might be oc's in the list that has
-	 * the OC_F_PURGED flag set. We do not want to let these slip through,
-	 * so we need to clear the flag before entering the do..while loop.
-	 */
-	Lck_Lock(&oh->mtx);
-	assert(oh->refcnt > 0);
-	VTAILQ_FOREACH(oc, &oh->objcs, hsh_list) {
-		CHECK_OBJ_NOTNULL(oc, OBJCORE_MAGIC);
-		assert(oc->objhead == oh);
-		oc->flags &= ~OC_F_PURGED;
-	}
-	Lck_Unlock(&oh->mtx);
 
-	do {
-		more = 0;
-		spc = ospc;
-		nobj = 0;
+	n_max = WS_ReserveLumps(wrk->aws, sizeof *ocp);
```

Comment:
We should have plenty of stack space here because of the de-facto minimum of [32k for pcree jit](https://pcre.org/current/doc/html/pcre2jit.html#stackcontrol), so we could either just use the stack, or at least use more stack elements as a fallback

Reply:
FWIW, `wrk->aws` _is_ the stack also...